R/fotmob_stats.R

```diff
@@ -33,49 +33,103 @@
     janitor::clean_names()
 }
 
-#' @importFrom tibble tibble
+.upper1 <- function(x) {
+  x <- tolower(x)
+  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
+  x
+}
+
+#' @importFrom tibble tibble as_tibble
 #' @importFrom rvest read_html html_elements html_attr
 #' @importFrom stringr str_replace str_replace str_detect
 #' @importFrom rlang maybe_missing
+#' @importFrom dplyr distinct arrange bind_rows
+#' @importFrom purrr imap_lgl keep map_dfr
 .fotmob_get_stat_and_season_options <- function(
   country,
   league_name,
   league_id,
   team_or_player,
+  page_url,
   cached
 ) {
+
   tables <- fotmob_get_league_tables(
     cached = cached,
     country = rlang::maybe_missing(country, NULL),
     league_name = rlang::maybe_missing(league_name, NULL),
     league_id = rlang::maybe_missing(league_id, NULL)
   )
-
   url <- sprintf(
     "https://www.fotmob.com%s/%ss",
     stringr::str_replace(tables$page_url[1], "overview", "stats"),
     team_or_player
   )
   page <- url %>% rvest::read_html()
-  hrefs <- page %>% rvest::html_elements(".SeeAllButton") %>% rvest::html_attr("href")
-  next_url <- sprintf(
-    "https://www.fotmob.com%s",
-    hrefs[1]
-  )
-  next_page <- next_url %>% rvest::read_html()
-  option_elements <- next_page %>% rvest::html_elements("option")
-  labels <- option_elements %>% rvest::html_text2()
-  rgx <- "(^.*)(\\s)(20[012].*$)"
-  league_name <- labels %>% stringr::str_replace(rgx, "\\1")
-  season_name <- labels %>% stringr::str_replace(rgx, "\\3")
-  league_name <- ifelse(season_name == league_name, NA_character_, league_name)
-  is_season <- season_name %>% stringr::str_detect("^2")
-  tibble::tibble(
-    league_name = league_name,
-    option_type = ifelse(is_season, "season", "stat"),
-    name = season_name,
-    id = option_elements %>% rvest::html_attr("value")
-  )
+  see_all_button <- page %>% rvest::html_elements(".SeeAllButton")
+  has_see_all_button <- length(see_all_button) > 0
```

Comment:
this is the big change with this PR.

1. Check if we're on a stats page with a "See All" button. This should exist during a league's season; but during the off-season, Fotmob may jump ahead to the next season, in which case there are no stats.
2. If there is a "See All" button, use the prior logic to scrape season and stat options.
3. If there isn't, then scrape the options from the JSON directly, e.g. https://www.fotmob.com/_next/data/3337/leagues/47/stats/premier-league.json